src/hotspot/share/cds/archiveBuilder.hpp

```diff
@@ -90,6 +90,18 @@ const int SharedSpaceObjectAlignment = KlassAlignmentInBytes;
 //    buffered_address + _buffer_to_requested_delta == requested_address
 //
 class ArchiveBuilder : public StackObj {
+public:
+  // Archived heap object headers carry pre-computed narrow Klass ids calculated with the
+  // following scheme:
+  // 1) the encoding base must be the mapping start address.
+  // 2) shift must be large enough to result in an encoding range that covers the runtime Klass range.
+  //    That Klass range is defined by CDS archive size and runtime class space size. Luckily, the maximum
+  //    size can be predicted: archive size is assumed to be <1G, class space size capped at 3G, and at
+  //    runtime we put both regions adjacent to each other. Therefore, runtime Klass range size < 4G.
+  //    Since nKlass itself is 32 bit, our encoding range len is 4G, and since we set the base directly
+  //    at mapping start, these 4G are enough. Therefore, we don't need to shift at all (shift=0).
```

Comment:
I would reformulate the first sentence. I spend a lot of cycles on the original text, and its not fully correct anymore.
Proposal:
```
The archive contains pre-computed narrow Klass IDs in two places:
- in the header of archived java objects (only if the archive contains java heap portions)
- within the prototype markword of archived Klass structures.
These narrow Klass ids have been computed at dump time with the following scheme:
blalblabla
```